### PR TITLE
Add an option to fail on duplicate key when parsing a json object

### PR DESCRIPTION
**Pull request by blitzkr1eg**
Fixes https://github.com/circe/circe/issues/1130
by adding a new flag to the parser: allowDuplicateKeys

Also added a factory to build parsers with custom parameters: ParserFactory

### CODE REVIEW

**Comment by blitzkr1eg** [comment]
I see the Travis build could not run. I don't know how to retrigger/fix.

**Comment by travisbrown** [comment]
We’ve been seeing these `openjdk8` failures for the past couple of days (but only on a couple of repos, I think—and restarts don’t seem to help). I’m traveling through Wednesday but will take a closer look when I’m back if nobody beats me to it.

File: modules/parser/jvm/src/main/scala/io/circe/parser/ParserFactory.scala [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package io.circe.parser
+
+import io.circe.Parser
+import io.circe.jawn.JawnParser
+
+case class ParserFactory(maxValueSize: Option[Int], allowDuplicateKeys: Boolean) {
```
Comment (travisbrown) [inline modules/parser/jvm/src/main/scala/io/circe/parser/ParserFactory.scala]: Why is this needed? I'd prefer to keep the Scala.js and JVM APIs in circe-parser identical.

File: modules/jawn/src/main/scala/io/circe/jawn/JawnParser.scala [hunk 2]
```diff
@@ -18,16 +18,37 @@ final object JawnParser {
...
+    * digits) may support denial-of-service attacks. For example, the string
+    * constructor for Java's `BigInteger` is quadratic with the length of the
+    * input, and decoding a ten-million digit JSON number into a `BigInteger` may
+    * take minutes.
+    */
+  def apply(maxValueSize: Int, allowDuplicateKeys: Boolean): JawnParser = new JawnParser(Some(maxValueSize), allowDuplicateKeys)
+
+
+  /**
+    * Returns a parser that fails on encountering duplicate keys as per JSONlint.
```
Comment (travisbrown) [inline modules/jawn/src/main/scala/io/circe/jawn/JawnParser.scala]: This comment doesn't seem accurate, since the user could provide `allowDuplicateKeys = true`.

File: modules/jawn/src/main/scala/io/circe/jawn/JawnParser.scala [hunk 3]
```diff
@@ -18,16 +18,37 @@ final object JawnParser {
...
+    */
+  def apply(allowDuplicateKeys: Boolean): JawnParser = new JawnParser(None, allowDuplicateKeys)
 }
 
-class JawnParser(maxValueSize: Option[Int]) extends Parser {
-  def this() = this(None)
+class JawnParser(maxValueSize: Option[Int], allowDuplicateKeys: Boolean) extends Parser {
```
Comment (travisbrown) [inline modules/jawn/src/main/scala/io/circe/jawn/JawnParser.scala]: Can you add a default `true` value here?

File: modules/jawn/src/main/scala/io/circe/jawn/CirceSupportParser.scala [hunk 4]
```diff
@@ -93,16 +91,24 @@ class CirceSupportParser(maxValueSize: Option[Int]) extends SupportParser[Json]
...
         key = null
       }
       final def finish(index: Int): Json = Json.fromJsonObject(JsonObject.fromLinkedHashMap(m))
       final def isObj: Boolean = true
     }
   }
 
+  private def safePut[K, V](key: K, value: V, map: LinkedHashMap[K, V], allowDuplicateKeys: Boolean): V = {
+    val oldValue = map.put(key, value)
+    if (oldValue != null && !allowDuplicateKeys)
```
Comment (travisbrown) [inline modules/jawn/src/main/scala/io/circe/jawn/CirceSupportParser.scala]: Can you use `oldValue.ne(null)`? Reference equality is fine for `null` checks.

File: modules/jawn/src/main/scala/io/circe/jawn/JawnParser.scala (hunk 2 above)
Comment (travisbrown) [reply modules/jawn/src/main/scala/io/circe/jawn/JawnParser.scala]: I think I'd prefer to omit this overload.

File: modules/jawn/src/main/scala/io/circe/jawn/CirceSupportParser.scala [hunk 5]
```diff
@@ -93,16 +91,24 @@ class CirceSupportParser(maxValueSize: Option[Int]) extends SupportParser[Json]
...
+      throw new IllegalArgumentException(s"Invalid json, duplicate key name found: $key")
```
Comment (travisbrown) [inline modules/jawn/src/main/scala/io/circe/jawn/CirceSupportParser.scala]: I think it'd be better to use `org.typelevel.jawn.ParseException` here, no?

Comment (travisbrown) [reply modules/jawn/src/main/scala/io/circe/jawn/CirceSupportParser.scala]: Nevermind—I agree `IllegalArgumentException` is the right thing to use here.

**Comment by travisbrown** [comment]
The commits here are included in #1149.
